### PR TITLE
Object storage driver

### PR DESCRIPTION
**Pull request by ntk148v**

### CODE REVIEW

File: calplus/conf/providers.py [hunk 1]
```diff
@@ -29,23 +29,25 @@
                default=''),
     cfg.StrOpt('os_novaclient_version',
                default='2.1'),
+    cfg.StrOpt('os_swiftclient_version',
+               default='2'),
```
Comment (hieulq) [inline calplus/conf/providers.py]: which OpenStack version you want to use? Newton is 3, so I think you want to use Mitaka, right?
https://releases.openstack.org/newton/#newton-python-swiftclient

File: calplus/tests/fake_config_file.conf [hunk 2]
```diff
@@ -20,6 +20,7 @@ os_identity_api_version = 3
 os_image_api_version = 2
 tenant_id =
 os_novaclient_version = 2.1
+os_swiftclient_version = 2
```
Comment (hieulq) [inline calplus/tests/fake_config_file.conf]: ditto

File: calplus/conf/providers.py [hunk 3]
```diff
@@ -29,23 +29,25 @@
                default=''),
     cfg.StrOpt('os_novaclient_version',
                default='2.1'),
```
Comment (hieulq) [inline calplus/conf/providers.py]: this should be changed along with your swift client version, too.
Maybe we should support Newton or Ocata?

File: calplus/tests/fake_config_file.conf [hunk 4]
```diff
@@ -20,6 +20,7 @@ os_identity_api_version = 3
 os_image_api_version = 2
 tenant_id =
 os_novaclient_version = 2.1
```
Comment (hieulq) [inline calplus/tests/fake_config_file.conf]: ditto

File: calplus/tests/fake_config_file.conf [hunk 5]
```diff
@@ -34,6 +35,7 @@ os_identity_api_version = 3
...
+os_swiftclient_version = 2
```
Comment (hieulq) [inline calplus/tests/fake_config_file.conf]: ditto

File: calplus/v1/network/drivers/base.py [hunk 6]
```diff
@@ -66,10 +67,10 @@ def get(self):
...
         # get all network quota from Cloud Provider.
         attrs = ("networks",
-                "security_groups",
-                "floating_ips",
-                "routers",
-                "internet_gateways")
+                 "security_groups",
+                 "floating_ips",
+                 "routers",
+                 "internet_gateways")
```
Comment (hieulq) [inline calplus/v1/network/drivers/base.py]: nice!

File: calplus/v1/object_storage/client.py [hunk 7]
```diff
@@ -1,14 +1,166 @@
...
     """Top-level object to access CAL API
     This class must be extended base.Singleton class to make
     sure only one instance of this one is ever created."""
 
-    def __init__(self, provider, cloud_config, *args, **kwargs):
+    def __init__(self, provider, cloud_config):
         BaseClient.__init__(self, CONF.object_storage.driver_path,
                             provider, cloud_config)
+
+    def create_container(self, container, **kwargs):
```
Comment (hieulq) [inline calplus/v1/object_storage/client.py]: the term container was used only in Swift context, how about S3?
maybe we should use the more generic term or write more details in docstring

File: calplus/v1/object_storage/drivers/amazon.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+            'aws_access_key_id': self.aws_access_key_id,
+            'aws_secret_access_key': self.aws_secret_access_key,
+            'region_name': self.region_name,
+            'endpoint_url': self.endpoint_url
+        }
+
+        self.client = boto3.client('s3', **parameters)
+        self.quota = AmazonQuota(self.client, self.limit)
+
+    def create_container(self, container, **kwargs):
```
Comment (hieulq) [inline calplus/v1/object_storage/drivers/amazon.py]: see, create container in S3 seems misunderstanding

File: calplus/v1/object_storage/drivers/base.py [hunk 9]
```diff
@@ -4,6 +4,133 @@
...
 @six.add_metaclass(abc.ABCMeta)
 class BaseDriver(object):
+    """The base class that all Driver classes should inherit from."""
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self):
+        super(BaseDriver, self).__init__()
+
+    @abc.abstractmethod
+    def create_container(self, container, **kwargs):
```
Comment (hieulq) [inline calplus/v1/object_storage/drivers/base.py]: ditto

File: calplus/v1/object_storage/drivers/openstack.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+                           password=self.password,
+                           project_domain_name=self.project_domain_name,
+                           project_name=self.project_name)
+        sess = session.Session(auth=auth)
+        self.client = Connection(self.client_version, session=sess)
+        self.quota = OpenStackQuota(
+            self.client, self.tenant_id, self.limit)
+
+    def create_container(self, container, **kwargs):
+        return self.client.put_container(container)
```
Comment (hieulq) [inline calplus/v1/object_storage/drivers/openstack.py]: overall LGTM, except the method name

File: requirements.txt [hunk 11]
```diff
@@ -9,3 +9,4 @@ python-neutronclient
 keystoneauth1
 boto3
 python-novaclient
+python-swiftclient
```
Comment (hieulq) [inline requirements.txt]: We should refactor this file for including specific range of version e.g: python-swiftclient >= 3.0.0

File: calplus/conf/providers.py (hunk 1 above)
Comment (ntk148v) [inline calplus/conf/providers.py]: Oh shit, i use wrong term here. this is not `os_swiftclient_version`. It should be `auth_version` - OpenStack auth version. Not like python-novaclient, python-swiftclient doesn't have several versions. 
I'll fix it ASAP.

File: requirements.txt (hunk 11 above)
Comment (ntk148v) [inline requirements.txt]: Ok boss.

File: calplus/tests/fake_config_file.conf (hunk 5 above)
Comment (hieulq) [inline calplus/tests/fake_config_file.conf]: how about this test case? auth or swiftclient

Comment (ntk148v) [inline calplus/tests/fake_config_file.conf]: forgot this, and i fix in [new commit](https://github.com/cloudcomputinghust/CAL/pull/51/commits/553d97e731eb6f771eb862431ad0273d89736d4d)

**Pull request merged**